### ISSUE TITLE
Problem: Verify Checksum “completed successfully” even though not valid

### ISSUE THREAD

**Issue by joel-simpson** [artefactual/archivematica#685]
**Verify Transfer Checksum “completed successfully” even though the checksum is not valid**
**Background:** testing in the demo-jisc environment for the Jisc RDSS project. I created a transfer with one file in it, which included an md5 checksum.  The checksum was provided in the ‘request preservation’ message that we mock from the rdss. This is sent using the msgCreator app to the Archivematica Channel Adaptor, which takes the checksum metadata (uuid, checksum type, checksum value) and creates a checksum file that is included in the transfer sent to Archivematica. 

I had already tested with a transfer where the checksum was valid, so in this case, I wanted to test what happened when the checksum provided was incorrect (i.e. would not match a new checksum generated by the same file). I changed 1 digit in the checksum value to do this.

**Expected Behaviour:** When the transfer was processed, I expect the Verify Transfer Checksum to fail; for that failure to trigger the “Failed Transfer” micro service; and then stop processing of the transfer.  

Actual Behaviour: The “Verify Transfer Checksum” indicated that it “completed successfully” and processing of the Transfer continued (straight through to ingest). 

a key detail: when I intentionally corrupted the checksum value, I replaced a hex character with a ’t’ (not considering that the checksum digits must all be hex characters). When I re-ran the test with a hex character instead of a t (but still not the original value), the micro service fails as expected. Normally we expect the checksum to be fine, but for the verification to fail because the file is corrupted. In my scenario, the file was fine but the checksum provided with it was corrupted. So this scenario may not happen frequently.

Output from the task is included below for reference. Discussed the defect with Jesús who wrote the channel adaptor code. His initial assessment is that this has to do with the verify checksum micro service, and is not due to the way the channel adaptor is creating creating a checksum file from metadata. 

— output Job Verify metadata directory checksums
```

Task 63716561-a312-4eb5-89d2-de615bc5a0f3

File UUID
Unknown
File name
test3 for checksum fail-a932afa6-e2db-4645-82df-529cc38cc3be
Client
31bf6692ac10_2
Exit code
0
Start time
6/27/2017, 6:15:25 PM
End time
6/27/2017, 6:15:25 PM
Created time
6/27/2017, 6:15:25 PM
Duration
< 1 seconds(s)
Module verifyMD5_v0.0
"%sharedPath%currentlyProcessing/test3 for checksum fail-a932afa6-e2db-4645-82df-529cc38cc3be/" "%date%" "%taskUUID%" "a932afa6-e2db-4645-82df-529cc38cc3be"
Standard streams

```
Standard output (stdout)
File Does not exist: /var/archivematica/sharedDirectory/currentlyProcessing/test3 for checksum fail-a932afa6-e2db-4645-82df-529cc38cc3be/metadata/checksum.sha1
File Does not exist: /var/archivematica/sharedDirectory/currentlyProcessing/test3 for checksum fail-a932afa6-e2db-4645-82df-529cc38cc3be/metadata/checksum.sha256
Standard error (stderr)
md5deep: /var/archivematica/sharedDirectory/currentlyProcessing/test3 for checksum fail-a932afa6-e2db-4645-82df-529cc38cc3be/metadata/checksum.md5: Unable to find any hashes in file, skipped.
md5deep: Unable to load any matching files.
Try `md5deep -h` for more information.
md5deep: /var/archivematica/sharedDirectory/currentlyProcessing/test3 for checksum fail-a932afa6-e2db-4645-82df-529cc38cc3be/metadata/checksum.md5: Unable to find any hashes in file, skipped.
md5deep: Unable to load any matching files.
Try `md5deep -h` for more information.

**Comment by jraddaoui** [comment]
Tested with a default AM Vagrant box, using the latest changes from the QA branches and the Verify Transfer Checksum task failed as it should. This what I did to try to reproduce:

- Started a transfer with the `~/archivematica-sampledata/TestTransfers/fixityCheckShouldFail` directory without modification. It failed in the Verify transfer checksums microservice.
- Entered the Vagrant box and fixed the checksum.md5 file:
```
vagrant ssh
cd ~/archivematica-sampledata/TestTransfers/fixityCheckShouldFail/objects
md5sum ./* > ../metadata/checksum.md5
```
- Fixed `metadata/checksum.md5`:
```
38ba8487812ef9d722f55253646ed713  ./Nemastylis_geminiflora_Flower.PNG
04f7802b45838fed393d45afadaa9dcc  ./oakland03.jp2
2eb15cb1834214b05d0083c691f9545f  ./WFPC01.GIF
```
- Started a transfer again from the same directory. It didn't fail:
![screenshot from 2017-08-09 14-44-23](https://user-images.githubusercontent.com/637040/29122090-52f22654-7d11-11e7-8ef7-6a2a9d722ebe.png)
- Modified the `metadata/checksum.md5` file changing the last char of the first hash to a 't'. Broken checksum file:
```
38ba8487812ef9d722f55253646ed71t  ./Nemastylis_geminiflora_Flower.PNG
04f7802b45838fed393d45afadaa9dcc  ./oakland03.jp2
2eb15cb1834214b05d0083c691f9545f  ./WFPC01.GIF
```
- Started a transfer again from the same directory. It failed (No hash found in line 1):
![screenshot from 2017-08-09 14-47-31](https://user-images.githubusercontent.com/637040/29122203-b5ac69f8-7d11-11e7-913c-83ade422ebc5.png)

@joel-simpson and @sevein, I'd love to know how is the checksum file added using the msgCreator app and the Archivematica Channel Adaptor. Could you show me how to do it or provide me with a test transfer?

**Comment by sevein** [comment]
@jraddaoui, did the transfer continue straight through ingest? I think what Joel is suggesting is that the workflow should be modified in order to interrupt the transfer.

**Comment by jraddaoui** [comment]
It was interrupted when it failed.

**Comment by sevein** [comment]
Sorry I missed that. If you edit one of the checksums in `fixityCheckShouldFail` to contain an invalid hash (e.g. `____8487812ef9d722f55253646ed713` instead of `38ba8487812ef9d722f55253646ed713`), does the microservice still fail?

**Comment by jraddaoui** [comment]
I did that replacing the last "3" by a "t", it failed, see the last screenshot. I could try tomorrow with the "_" but I think it will fail with the same error.

**Comment by sevein** [comment]
Right, sorry, I can't read. `t` is definitely an invalid hex symbol :)
I'll give it a try. Thanks.

**Comment by sevein** [comment]
I can reproduce only when the file doesn't contain any valid checksum, e.g.:

```
____5cb1834214b05d0083c691f9545f  ./WFPC01.GIF
____802b45838fed393d45afadaa9dcc  ./oakland03.jp2
____9676ead317f643e9a58f0177d1e6  ./Nemastylis_geminiflora_Flower.PNG
```

This results in the behavior described by Joel.

**Comment by jraddaoui** [comment]
Wow, that's rare. Thanks @sevein

**Comment by jraddaoui** [comment]
https://github.com/artefactual/archivematica/pull/711

**Comment by sevein** [comment]
Fixed in #711.

**Comment by sallain** [comment]
Tested this on qa/1.x (to be 1.7) with the Archivematica sample data transfer TestTransfers/fixityCheckShouldFail and it failed as expected.
